aggregator/src/aggregator/key_rotator.rs

```diff
@@ -0,0 +1,673 @@
+        {
+            Err(DatastoreError::User(
+                anyhow!("unexpected state: no keypairs are active").into(),
+            ))
+        } else {
+            Ok(())
+        }
+    }
+
+    #[tracing::instrument(skip(tx, current_keypairs, available_ids))]
+    async fn run_hpke_for_config(
+        tx: &Transaction<'_, C>,
+        current_keypairs: &[GlobalHpkeKeypair],
+        available_ids: &mut impl Iterator<Item = HpkeConfigId>,
+        config: &HpkeKeyRotatorConfig,
+    ) -> Result<(), DatastoreError> {
+        let (pending_keypairs, active_keypairs, expired_keypairs) =
+            Self::partition_keypairs(current_keypairs, &config.ciphersuite);
+        let mut next_key = || {
+            generate_hpke_config_and_private_key(
+                available_ids
+                    .next()
+                    .ok_or_else(|| DatastoreError::User(anyhow!("u8 space exhausted").into()))?,
+                config.ciphersuite.kem_id(),
+                config.ciphersuite.kdf_id(),
+                config.ciphersuite.aead_id(),
+            )
+            .map_err(|e| DatastoreError::User(e.into()))
+        };
+
+        if active_keypairs.is_empty() && pending_keypairs.is_empty() {
+            // Bootstrapping case: there are no keys at all for this ciphersuite, so we need to
+            // insert one.
+            let keypair = next_key()?;
+            info!(
+                id = ?keypair.config().id(),
+                "bootstrapping: inserting key for new ciphersuite"
+            );
+            tx.put_global_hpke_keypair(&keypair).await?;
+
+            // If there are zero keypairs reported by the database, it's likely beacuse we're
+            // in a brand new database, and this is the first execution of Janus. Initialize
+            // the table with active keys. Janus won't be ready until there's at least one
+            // active keypair in the database.
+            //
+            // Otherwise, the ciphersuite was likely added to the configuration and its key
+            // needs to go through the normal pending->active state change.
+            if current_keypairs.is_empty() {
+                info!(id = ?keypair.config().id(), "bootstrapping: moving new key to active state");
+                tx.set_global_hpke_keypair_state(keypair.config().id(), &HpkeKeyState::Active)
+                    .await?;
+            }
+        } else {
+            let to_be_expired_keypairs: Vec<_> = active_keypairs
+                .iter()
+                .filter(|keypair| {
+                    Self::duration_since(tx, keypair.updated_at()) > config.active_duration
+                })
+                .cloned()
+                .collect();
+
+            if to_be_expired_keypairs != active_keypairs {
+                for keypair in to_be_expired_keypairs {
+                    info!(id = ?keypair.id(), "allowing already active keypair to replace this one");
```

Comment:
IMO this message should be a bit more specific that the relevant key's state is being set to expired.
```suggestion
                    info!(id = ?keypair.id(), "multiple active keys are present, marking key expired");
```